### PR TITLE
Refactor dropout cuda impl for code reuse.

### PR DESCRIPTION
**Pull request by limin2021**
### PR types
Performance optimization

### PR changes
OPs
### Describe
Refactor dropout cuda impl for code reuse:
(1) Move dropout cuda forward impl in dropout_op.cu and backward impl in dropout_op.h to newly dropout_impl.cu.h;
(2) Originally, the backward impl for cpu and gpu share same code. To be more clear, they are separated into two pieces.
I extract the cuda backward impl from dropout_op.h and put them in dropout_impl.cu.h.
Now, all the cpu impl is in dropout_op.h and all the gpu impl is in dropout_impl.cu.h.

Unittest Results for dropout op:
c49cf754fcd67afc707fe39becc4d626

### CODE REVIEW

**Comment by xingfeng01** [comment]
建议 Describe 里第 (2) 条描述稍微的详细一点。

**Comment by xingfeng01** [comment]
LGTM

**Comment by limin2021** [comment]
> 建议 Describe 里第 (2) 条描述稍微的详细一点。

Done.

File: paddle/fluid/operators/dropout_impl.cu.h [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+          dest_vec[ii] = src_vec[ii];
+        }
+        mask_vec[ii] = 1;
+      }
+    }
+
+    *(reinterpret_cast<LoadT*>(&dst[i])) =
+        *reinterpret_cast<LoadT*>(&dest_vec[0]);
+    *(reinterpret_cast<MaskLoadT*>(&mask_data[i])) =
+        *reinterpret_cast<MaskLoadT*>(&mask_vec[0]);
```
Comment (zhangting2020) [inline paddle/fluid/operators/dropout_impl.cu.h]: 这里在当前的devleop已经不是这样写了。注意下挪动后，不要修改了内容

Comment (limin2021) [reply paddle/fluid/operators/dropout_impl.cu.h]: Done.

**Review by zhangting2020 — APPROVED** [review]

**Pull request merged**